File: shogi-ai/app/usi_engine.py
"""
USIエンジンラッパークラス
やねうら王との通信を管理
"""

import asyncio
import time
import logging
from typing import Optional, List, Dict
from asyncio import Queue

logger = logging.getLogger(__name__)

class USIEngine:
    """単一のUSIエンジンインスタンス（非同期対応）"""

    def __init__(self, engine_path: str):
        self.engine_path = engine_path
        self.process: Optional[asyncio.subprocess.Process] = None
        self.output_queue: Queue = Queue()
        self._reader_task: Optional[asyncio.Task] = None
        self._running = False
        self.engine_info: Dict[str, str] = {}

    async def start(self):
        """エンジンプロセスを起動"""
        self.process = await asyncio.create_subprocess_exec(
            self.engine_path,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            cwd="/engine"  # 評価関数ファイルのあるディレクトリで起動
        )
        self._running = True
        self._reader_task = asyncio.create_task(self._read_output())
        logger.info(f"Engine started: {self.engine_path}")

    async def _read_output(self):
        """エンジンの出力を非同期で読み込む"""
        while self._running and self.process and self.process.stdout:
            try:
                line = await self.process.stdout.readline()
                if not line:
                    break
                decoded = line.decode('utf-8', errors='ignore').strip()
                if decoded:
                    await self.output_queue.put(decoded)
            except Exception as e:
                logger.error(f"Error reading engine output: {e}")
                break

    async def send_command(self, command: str):
        """エンジンにコマンドを送信"""
        if self.process and self.process.stdin:
            self.process.stdin.write(f"{command}\n".encode('utf-8'))
            await self.process.stdin.drain()
            logger.debug(f"Sent: {command}")

    async def wait_for_response(self, expected: str, timeout: float = 5.0) -> bool:
        """特定のレスポンスを待機"""
        start_time = time.time()
        temp_buffer = []

        while time.time() - start_time < timeout:
            try:
                line = await asyncio.wait_for(self.output_queue.get(), timeout=0.5)
                logger.debug(f"Received: {line}")

                if line == expected:
                    # 一時バッファをキューに戻す
                    for buffered_line in temp_buffer:
                        await self.output_queue.put(buffered_line)
                    return True

                temp_buffer.append(line)

            except asyncio.TimeoutError:
                continue

        # タイムアウト時にバッファを戻す
        for buffered_line in temp_buffer:
            await self.output_queue.put(buffered_line)
        return False

    async def initialize(self) -> Dict[str, str]:
        """エンジンを初期化してusi, isreadyを送信"""
        await self.send_command("usi")

        start_time = time.time()
        while time.time() - start_time < 10.0:
            try:
                line = await asyncio.wait_for(self.output_queue.get(), timeout=0.5)
                logger.debug(f"Init: {line}")

                if line.startswith("id name"):
                    self.engine_info["name"] = line[8:].strip()
                elif line.startswith("id author"):
                    self.engine_info["author"] = line[10:].strip()
                elif line == "usiok":
                    logger.info(f"Engine initialized: {self.engine_info.get('name', 'Unknown')}")
                    break

            except asyncio.TimeoutError:
                continue

        # isreadyで準備完了を待つ
        await self.send_command("isready")
        await self.wait_for_response("readyok", timeout=10.0)

        return self.engine_info

    async def set_option(self, name: str, value: str):
        """エンジンオプションを設定"""
        await self.send_command(f"setoption name {name} value {value}")
        await asyncio.sleep(0.05)  # オプション設定の反映待ち

    async def set_position(self, sfen: Optional[str] = None, moves: Optional[List[str]] = None):
        """局面を設定"""
        if sfen:
            cmd = f"position sfen {sfen}"
        else:
            cmd = "position startpos"

        if moves:
            cmd += " moves " + " ".join(moves)

        await self.send_command(cmd)

    async def go_analyze(self, time_ms: Optional[int] = None, depth: Optional[int] = None,
                        multipv: int = 1) -> Dict:
        """解析を実行してbestmoveと情報を取得"""

        # MultiPV設定
        if multipv > 1:
            await self.set_option("MultiPV", str(multipv))

        # goコマンド構築
        cmd = "go"
        if depth:
            cmd += f" depth {depth}"
        elif time_ms:
            cmd += f" byoyomi {time_ms}"
        else:
            cmd += " byoyomi 1000"

        await self.send_command(cmd)

        # info行を収集
        variations = {}  # {multipv_num: {...}}
        bestmove = None
        start_time = time.time()

        while True:
            try:
                line = await asyncio.wait_for(
                    self.output_queue.get(),
                    timeout=max(30.0, (time_ms or 1000) / 1000 + 5.0)
                )
                logger.debug(f"Go: {line}")

                if line.startswith("info"):
                    info = self._parse_info_line(line)
                    if info and "multipv" in info:
                        pv_num = info["multipv"]
                        variations[pv_num] = info

                elif line.startswith("bestmove"):
                    parts = line.split()
                    if len(parts) >= 2:
                        bestmove = parts[1]
                    break

            except asyncio.TimeoutError:
                logger.warning("Go command timeout")
                break

        elapsed_ms = int((time.time() - start_time) * 1000)

        return {
            "bestmove": bestmove or "resign",
            "variations": [variations[i] for i in sorted(variations.keys())],
            "time_ms": elapsed_ms
        }

    async def go_mate(self, time_ms: int) -> Dict:
        """詰み探索を実行"""
        await self.send_command(f"go mate {time_ms}")

        start_time = time.time()

        while True:
            try:
                line = await asyncio.wait_for(
                    self.output_queue.get(),
                    timeout=(time_ms / 1000 + 5.0)
                )
                logger.debug(f"Mate: {line}")

                if line.startswith("checkmate"):
                    parts = line.split()

                    if len(parts) == 1 or parts[1] == "nomate":
                        return {
                            "mate_found": False,
                            "message": "No mate found"
                        }
                    elif parts[1] == "timeout":
                        return {
                            "mate_found": False,
                            "message": "Search timeout"
                        }
                    elif parts[1] == "notimplemented":
                        return {
                            "mate_found": False,
                            "message": "Mate search not implemented"
                        }
                    else:
                        # 詰み手順
                        mate_moves = parts[1:]
                        return {
                            "mate_found": True,
                            "mate_moves": mate_moves,
                            "mate_length": len(mate_moves)
                        }

            except asyncio.TimeoutError:
                return {
                    "mate_found": False,
                    "message": "Mate search timeout"
                }

    def _parse_info_line(self, line: str) -> Optional[Dict]:
        """info行をパース"""
        parts = line.split()
        if len(parts) < 2 or parts[0] != "info":
            return None

        info = {}
        i = 1

        while i < len(parts):
            if parts[i] == "depth" and i + 1 < len(parts):
                info["depth"] = int(parts[i + 1])
                i += 2
            elif parts[i] == "score" and i + 2 < len(parts):
                score_type = parts[i + 1]
                score_value = int(parts[i + 2])
                if score_type == "cp":
                    info["score_cp"] = score_value
                elif score_type == "mate":
                    info["score_mate"] = score_value
                i += 3
            elif parts[i] == "nodes" and i + 1 < len(parts):
                info["nodes"] = int(parts[i + 1])
                i += 2
            elif parts[i] == "multipv" and i + 1 < len(parts):
                info["multipv"] = int(parts[i + 1])
                i += 2
            elif parts[i] == "pv":
                info["pv"] = parts[i + 1:]
                break
            else:
                i += 1

        return info if info else None

    async def quit(self):
        """エンジンを終了"""
        self._running = False
        await self.send_command("quit")

        if self.process:
            try:
                await asyncio.wait_for(self.process.wait(), timeout=3.0)
            except asyncio.TimeoutError:
                self.process.kill()
                await self.process.wait()

        if self._reader_task:
            self._reader_task.cancel()
            try:
                await self._reader_task
            except asyncio.CancelledError:
                pass

        logger.info("Engine terminated")


class USIEnginePool:
    """USIエンジンのプール（複数リクエスト対応）"""

    def __init__(self, engine_path: str, pool_size: int = 4, options: Optional[Dict[str, str]] = None):
        self.engine_path = engine_path
        self.pool_size = pool_size
        self.options = options or {}
        self.engines: List[USIEngine] = []
        self.available: Queue = Queue()
        self._engine_info: Dict[str, str] = {}

    async def initialize(self):
        """エンジンプールを初期化"""
        for i in range(self.pool_size):
            engine = USIEngine(self.engine_path)
            await engine.start()
            info = await engine.initialize()

            if i == 0:
                self._engine_info = info

            # オプション設定
            for name, value in self.options.items():
                await engine.set_option(name, value)

            self.engines.append(engine)
            await self.available.put(engine)

        logger.info(f"Engine pool initialized with {self.pool_size} engines")

    async def acquire(self) -> USIEngine:
        """エンジンを取得"""
        return await self.available.get()

    async def release(self, engine: USIEngine):
        """エンジンを返却"""
        # usinewgameで状態をリセット
        await engine.send_command("usinewgame")
        await engine.send_command("isready")
        await engine.wait_for_response("readyok", timeout=5.0)

        await self.available.put(engine)

    async def analyze(self, sfen: Optional[str] = None, moves: Optional[List[str]] = None,
                     time_ms: int = 1000, depth: Optional[int] = None,
                     multipv: int = 1) -> Dict:
        """局面を解析"""
        engine = await self.acquire()

        try:
            await engine.set_position(sfen=sfen, moves=moves)
            result = await engine.go_analyze(time_ms=time_ms, depth=depth, multipv=multipv)

            # レスポンス整形
            variations = []
            for var in result["variations"]:
                variations.append({
                    "move": var.get("pv", [""])[0] if var.get("pv") else "",
                    "score_cp": var.get("score_cp"),
                    "score_mate": var.get("score_mate"),
                    "depth": var.get("depth", 0),
                    "nodes": var.get("nodes"),
                    "pv": var.get("pv")
                })

            return {
                "bestmove": result["bestmove"],
                "variations": variations,
                "time_ms": result["time_ms"],
                "engine_name": self._engine_info.get("name", "YaneuraOu")
            }

        finally:
            await self.release(engine)

    async def search_mate(self, sfen: Optional[str] = None, moves: Optional[List[str]] = None,
                         time_ms: int = 5000) -> Dict:
        """詰み探索"""
        engine = await self.acquire()

        try:
            await engine.set_position(sfen=sfen, moves=moves)
            result = await engine.go_mate(time_ms=time_ms)
            return result

        finally:
            await self.release(engine)

    def is_healthy(self) -> bool:
        """プールの健全性チェック"""
        return len(self.engines) > 0 and all(e.process and e.process.returncode is None for e in self.engines)

    def get_engine_info(self) -> Dict:
        """エンジン情報を取得"""
        return {
            "name": self._engine_info.get("name", "YaneuraOu"),
            "author": self._engine_info.get("author", "Unknown"),
            "version": "latest",
            "options": self.options
        }

    async def shutdown(self):
        """全エンジンを終了"""
        for engine in self.engines:
            await engine.quit()
        logger.info("Engine pool shutdown complete")

File: server/src/lib/shogiAi/usiEngine.ts
import { AsyncQueue } from "./asyncQueue";
import type { AnalyzeResponse, EngineInfo, MateSearchResponse, MoveInfo } from "./types";

/**
 * USIエンジンとの通信を管理する単一エンジンインスタンス
 *
 * Bun.spawnでYaneuraOuプロセスを起動し、stdin/stdoutをパイプで接続する。
 * バックグラウンドで標準出力を読み続けて行単位でキューに積み、コマンドごとに必要な行を取り出す。
 */
export class UsiEngine {
  private readonly enginePath: string;
  private readonly engineCwd: string;
  private readonly outputQueue = new AsyncQueue<string>();
  private process: ReturnType<typeof Bun.spawn> | null = null;
  private readerPromise: Promise<void> | null = null;
  private running = false;
  private info: EngineInfo = {
    name: "YaneuraOu",
    author: "Unknown",
    version: "latest",
    options: {},
  };

  constructor(params: { enginePath: string; engineCwd: string }) {
    this.enginePath = params.enginePath;
    this.engineCwd = params.engineCwd;
  }

  /** エンジンプロセスを起動する */
  async start(): Promise<void> {
    this.process = Bun.spawn([this.enginePath], {
      stdin: "pipe",
      stdout: "pipe",
      stderr: "pipe",
      cwd: this.engineCwd,
    });
    this.running = true;
    this.readerPromise = this.readOutputLoop();
  }

  /** エンジンの出力を非同期に読み込み、行単位でキューに積む */
  private async readOutputLoop(): Promise<void> {
    if (!this.process?.stdout) return;

    const reader = (this.process.stdout as ReadableStream<Uint8Array>).getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    try {
      while (this.running) {
        const { value, done } = await reader.read();
        if (done) break;
        buffer += decoder.decode(value, { stream: true });

        let newlineIdx: number;
        while ((newlineIdx = buffer.indexOf("\n")) >= 0) {
          const line = buffer.slice(0, newlineIdx).trim();
          buffer = buffer.slice(newlineIdx + 1);
          if (line.length > 0) {
            this.outputQueue.push(line);
          }
        }
      }
    } catch {
      // プロセス終了時は握りつぶす
    } finally {
      reader.releaseLock();
      this.outputQueue.close();
    }
  }

  /** エンジンにコマンドを1行送信する */
  async sendCommand(command: string): Promise<void> {
    const stdin = this.process?.stdin;
    if (!stdin || typeof stdin === "number") {
      throw new Error("Engine process stdin is not available");
    }
    stdin.write(`${command}\n`);
    await stdin.flush();
  }

  /**
   * 指定した応答が来るまで待機する
   *
   * @param expected - 期待する1行の文字列（完全一致）
   * @param timeoutMs - タイムアウト（ミリ秒）
   * @returns 応答を受け取ったらtrue
   */
  async waitForResponse(expected: string, timeoutMs: number): Promise<boolean> {
    const start = Date.now();
    const buffered: string[] = [];

    while (Date.now() - start < timeoutMs) {
      const line = await this.outputQueue.get(500);
      if (line === null) continue;

      if (line === expected) {
        for (const b of buffered) this.outputQueue.push(b);
        return true;
      }
      buffered.push(line);
    }

    for (const b of buffered) this.outputQueue.push(b);
    return false;
  }

  /** usi, isready を送信してエンジンを初期化する */
  async initialize(): Promise<EngineInfo> {
    await this.sendCommand("usi");

    const infoDeadline = Date.now() + 10_000;
    while (Date.now() < infoDeadline) {
      const line = await this.outputQueue.get(500);
      if (line === null) continue;

      if (line.startsWith("id name")) {
        this.info = { ...this.info, name: line.slice(8).trim() };
      } else if (line.startsWith("id author")) {
        this.info = { ...this.info, author: line.slice(10).trim() };
      } else if (line === "usiok") {
        break;
      }
    }

    await this.sendCommand("isready");
    await this.waitForResponse("readyok", 10_000);
    return this.info;
  }

  /** エンジンオプションを設定する */
  async setOption(name: string, value: string): Promise<void> {
    await this.sendCommand(`setoption name ${name} value ${value}`);
    await new Promise((resolve) => setTimeout(resolve, 50));
  }

  /** 局面を設定する */
  async setPosition(params: { sfen?: string | null; moves?: string[] | null }): Promise<void> {
    const { sfen, moves } = params;
    let cmd = sfen ? `position sfen ${sfen}` : "position startpos";
    if (moves && moves.length > 0) {
      cmd += ` moves ${moves.join(" ")}`;
    }
    await this.sendCommand(cmd);
  }

  /**
   * 局面解析を実行してbestmoveと候補手情報を取得する
   */
  async goAnalyze(params: {
    timeMs?: number | null;
    depth?: number | null;
    multipv: number;
  }): Promise<AnalyzeResponse> {
    const { timeMs, depth, multipv } = params;

    // MultiPVは状態が残るので常に明示的に設定する
    await this.setOption("MultiPV", String(multipv));

    let cmd = "go";
    if (depth) {
      cmd += ` depth ${depth}`;
    } else if (timeMs) {
      cmd += ` byoyomi ${timeMs}`;
    } else {
      cmd += " byoyomi 1000";
    }
    await this.sendCommand(cmd);

    const variations = new Map<number, MoveInfo>();
    let bestmove: string | null = null;
    const startedAt = Date.now();
    const readTimeoutMs = Math.max(30_000, (timeMs ?? 1000) + 5_000);
    const deadline = startedAt + readTimeoutMs;

    while (Date.now() < deadline) {
      const line = await this.outputQueue.get(500);
      if (line === null) continue;

      if (line.startsWith("info")) {
        const info = parseInfoLine(line);
        if (info && info.multipv !== undefined) {
          variations.set(info.multipv, info.move);
        }
      } else if (line.startsWith("bestmove")) {
        const parts = line.split(/\s+/);
        bestmove = parts[1] ?? null;
        break;
      }
    }

    const elapsedMs = Date.now() - startedAt;
    const sortedVariations = [...variations.entries()].sort(([a], [b]) => a - b).map(([, v]) => v);

    return {
      bestmove: bestmove ?? "resign",
      variations: sortedVariations,
      timeMs: elapsedMs,
      engineName: this.info.name,
    };
  }

  /** 詰み探索を実行する */
  async goMate(timeMs: number): Promise<MateSearchResponse> {
    await this.sendCommand(`go mate ${timeMs}`);

    const deadline = Date.now() + timeMs + 5_000;

    while (Date.now() < deadline) {
      const line = await this.outputQueue.get(500);
      if (line === null) continue;

      if (line.startsWith("checkmate")) {
        const parts = line.split(/\s+/);

        if (parts.length === 1 || parts[1] === "nomate") {
          return { mateFound: false, message: "No mate found" };
        }
        if (parts[1] === "timeout") {
          return { mateFound: false, message: "Search timeout" };
        }
        if (parts[1] === "notimplemented") {
          return { mateFound: false, message: "Mate search not implemented" };
        }

        const mateMoves = parts.slice(1);
        return {
          mateFound: true,
          mateMoves,
          mateLength: mateMoves.length,
        };
      }
    }

    return { mateFound: false, message: "Mate search timeout" };
  }

  /** エンジンを終了する */
  async quit(): Promise<void> {
    this.running = false;
    try {
      await this.sendCommand("quit");
    } catch {
      // 既に閉じている場合は無視
    }

    if (this.process) {
      const exitPromise = this.process.exited;
      const timeoutPromise = new Promise<void>((resolve) => setTimeout(resolve, 3_000));
      await Promise.race([exitPromise, timeoutPromise]);

      if (this.process.exitCode === null) {
        this.process.kill();
        await this.process.exited;
      }
    }

    if (this.readerPromise) {
      await this.readerPromise.catch(() => undefined);
    }
  }

  /** プロセスが生きているか */
  isAlive(): boolean {
    return this.process !== null && this.process.exitCode === null;
  }

  /** エンジン情報を取得する */
  getInfo(): EngineInfo {
    return { ...this.info };
  }
}

interface ParsedInfo {
  multipv?: number;
  move: MoveInfo;
}

/** USIのinfo行をパースする */
function parseInfoLine(line: string): ParsedInfo | null {
  const parts = line.split(/\s+/);
  if (parts.length < 2 || parts[0] !== "info") return null;

  let depth = 0;
  let scoreCp: number | null = null;
  let scoreMate: number | null = null;
  let nodes: number | null = null;
  let multipv: number | undefined;
  let pv: string[] | null = null;

  let i = 1;
  while (i < parts.length) {
    const token = parts[i];
    if (token === "depth" && i + 1 < parts.length) {
      depth = Number.parseInt(parts[i + 1] ?? "0", 10);
      i += 2;
    } else if (token === "score" && i + 2 < parts.length) {
      const type = parts[i + 1];
      const value = Number.parseInt(parts[i + 2] ?? "0", 10);
      if (type === "cp") scoreCp = value;
      else if (type === "mate") scoreMate = value;
      i += 3;
    } else if (token === "nodes" && i + 1 < parts.length) {
      nodes = Number.parseInt(parts[i + 1] ?? "0", 10);
      i += 2;
    } else if (token === "multipv" && i + 1 < parts.length) {
      multipv = Number.parseInt(parts[i + 1] ?? "1", 10);
      i += 2;
    } else if (token === "pv") {
      pv = parts.slice(i + 1);
      break;
    } else {
      i += 1;
    }
  }

  const move = pv && pv.length > 0 ? (pv[0] ?? "") : "";

  return {
    multipv,
    move: {
      move,
      scoreCp,
      scoreMate,
      depth,
      nodes,
      pv,
    },
  };
}
